src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs

```diff
@@ -41,14 +42,63 @@ public static SocketError GetLastSocketError()
+            IntPtr handle = Interop.Winsock.WSASocketW(addressFamily, socketType, protocolType, IntPtr.Zero, 0, Interop.Winsock.SocketConstructorFlags.WSA_FLAG_OVERLAPPED |
+                                                                                                                Interop.Winsock.SocketConstructorFlags.WSA_FLAG_NO_HANDLE_INHERIT);
 
             socket = new SafeSocketHandle(handle, ownsHandle: true);
             if (NetEventSource.IsEnabled) NetEventSource.Info(null, socket);
 
             return socket.IsInvalid ? GetLastSocketError() : SocketError.Success;
         }
 
+        public static unsafe SocketError CreateSocket(
+            SocketInformation socketInformation,
+            out SafeSocketHandle socket,
+            ref AddressFamily addressFamily,
+            ref SocketType socketType,
+            ref ProtocolType protocolType)
+        {
+            if (socketInformation.ProtocolInformation == null || socketInformation.ProtocolInformation.Length < Interop.Winsock.WSAProtocolInfo.Size)
+            {
+                throw new ArgumentException(SR.net_sockets_invalid_socketinformation, nameof(socketInformation));
+            }
+
+            fixed (byte* pinnedProtocolInformation = socketInformation.ProtocolInformation)
+            {
+                // Sockets are non-inheritable in .NET Core.
+                // Handle properties like HANDLE_FLAG_INHERIT are not cloned with socket duplication, therefore
+                // we need to disable handle inheritance when constructing the new socket handle from Protcol Info.
+                // Additionally, it looks like WSA_FLAG_NO_HANDLE_INHERIT has no effect when being used with the Protocol Info
+                // variant of WSASocketW, so it is being passed to that call only for consistency.
+                // Inheritance is being disabled with SetHandleInformation(...) after the WSASocketW call.
+                IntPtr handle = Interop.Winsock.WSASocketW(
+                    (AddressFamily)(-1),
+                    (SocketType)(-1),
+                    (ProtocolType)(-1),
+                    (IntPtr)pinnedProtocolInformation,
+                    0,
+                    Interop.Winsock.SocketConstructorFlags.WSA_FLAG_OVERLAPPED |
+                    Interop.Winsock.SocketConstructorFlags.WSA_FLAG_NO_HANDLE_INHERIT);
+
+                socket = new SafeSocketHandle(handle, ownsHandle: true);
+                if (NetEventSource.IsEnabled) NetEventSource.Info(null, socket);
+
+                if (socket.IsInvalid)
+                {
+                    return GetLastSocketError();
+                }
+
+                Interop.Kernel32.SetHandleInformation(socket, Interop.Kernel32.HandleFlags.Inherit, 0);
+
+                Interop.Winsock.WSAProtocolInfo protocolInfo = Marshal.PtrToStructure<Interop.Winsock.WSAProtocolInfo>((IntPtr)pinnedProtocolInformation);
```

Comment:
We can avoid using PtrToStructure once the previously discussed changes are made around making the struct blittable.